src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,256 @@
+        {
+            var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(operationBlockStartAnalysisContext.Compilation);
+
+            var oneParameterDeferredMethods = GetOneParameterDeferredMethods(wellKnownTypeProvider);
+            var twoParametersDeferredMethods = GetTwoParametersDeferredMethods(wellKnownTypeProvider);
+
+            var oneParameterEnumeratedMethods = GetOneParameterEnumeratedMethods(wellKnownTypeProvider);
+            var twoParametersEnumeratedMethods = GetTwoParametersEnumeratedMethods(wellKnownTypeProvider);
+
+            var compilation = operationBlockStartAnalysisContext.Compilation;
+            var additionalDeferTypes = GetTypes(compilation, s_additionalDeferredTypes);
+
+            var wellKnownSymbolsInfo = new WellKnownSymbolsInfo(
+                oneParameterDeferredMethods,
+                twoParametersDeferredMethods,
+                oneParameterEnumeratedMethods,
+                twoParametersEnumeratedMethods,
+                additionalDeferTypes);
+
+            var potentialDiagnosticOperationsBuilder = ImmutableHashSet.CreateBuilder<IOperation>();
+            operationBlockStartAnalysisContext.RegisterOperationAction(
+                context => CollectPotentialDiagnosticOperations(
+                    context,
+                    wellKnownSymbolsInfo,
+                    potentialDiagnosticOperationsBuilder),
+                OperationKind.ParameterReference,
+                OperationKind.LocalReference);
+
+            operationBlockStartAnalysisContext.RegisterOperationBlockEndAction(
+                context => Analyze(
+                    context,
+                    wellKnownTypeProvider,
+                    wellKnownSymbolsInfo,
+                    potentialDiagnosticOperationsBuilder));
+        }
+
+        private static void CollectPotentialDiagnosticOperations(
+            OperationAnalysisContext context,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            ImmutableHashSet<IOperation>.Builder builder)
+        {
+            var operation = context.Operation;
+            if (IsDeferredType(operation.Type.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                var isEnumerated = IsOperationEnumeratedByMethodInvocation(operation, wellKnownSymbolsInfo) || IsOperationEnumeratedByForEachLoop(operation, wellKnownSymbolsInfo);
+                if (isEnumerated)
+                    builder.Add(operation);
+            }
+        }
+
+        private void Analyze(
+            OperationBlockAnalysisContext context,
+            WellKnownTypeProvider wellKnownTypeProvider,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            ImmutableHashSet<IOperation>.Builder potentialDiagnosticOperationsBuilder)
+        {
+            var potentialDiagnosticOperations = potentialDiagnosticOperationsBuilder.ToImmutable();
+            if (potentialDiagnosticOperations.IsEmpty)
```

Comment:
Should this check be `potentialDiagnosticOperations.Count < 2` as we only care about multiple enumeration?

Reply:
Emm, this might be true for this PR. ( I am working on another PR that fixes some cases that could happen when Count < 2, but I can do that in that PR)